### DISCUSSION TITLE
GLTFExporter and Scene Viewer don't get along

### DISCUSSION THREAD

**Discussion by yoavniran** [google/model-viewer#4704] [Q&A]
Hi there, 
I'm guessing this is a long shot, but perhaps someone has some experience with the following:

I'm working on a `3D Viewer / AR Viewer` that also allows users to apply color/textures changes on the viewed model.
Then I want to make it possible to see the model in AR on a mobile device **including** the user made changes.

I was inspired by the iOS Quick Look [code](https://github.com/google/model-viewer/blob/master/packages/model-viewer/src/features/ar.ts#L437) to use the GLTFExporter in order to export the modified scene on the fly and hand it over to the SceneViewer as an Object URL. However, the scene viewer always opens and then closes (crashes) without being able to load the model.

With several models I tried, even if I just run the model through the export, without any changes, it just seems to break.

In the case of the Astronaut, not making changes and running through the exporter did load fine in SV. However, if I change the color of a material and try to export and load into the Viewer, it **breaks**. The model-viewer editor is able to show it just fine though:

<img width="420" alt="image" src="https://github.com/google/model-viewer/assets/1102278/ff0156c2-4c8e-4719-8407-1d347134b135">

Is there a known issue with the Exporter? Or is there a way to debug why the SV is crashing?

I tried playing with the exporter options (trs, onlyVisible, includeCustomExtensions) but that didn't seem to have any effect. 

This is my (simplified) code: 

```javascript

const exportGltf = (options) => {
					const exporter = new GLTFExporter();
					return new Promise((resolve, reject) => {
						exporter.parse(gltf.scene, resolve, reject, options);
					});
				},


const loadAr = async () => {
	const anchor = document.createElement("a");

	const buffer = await exportGltf();
	const blob = new Blob([buffer], {
		type: "model/gltf-binary",
	});

	const objectUrl = URL.createObjectURL(blob);
	const exportedUrl = new URL(objectUrl, self.location.toString());

       anchor.setAttribute("href",
                    getSceneViewerIntentUrl({
			    src: exportedUrl,
			    fallbackUrl,
			}));

      anchor.click();
};
```

Thanks!

**Comment by elalish** [comment]
I think you have a misunderstanding - Scene Viewer is Android-only; you mention iOS, which would be Quick Look. Also, Quick Look can represent an updated model because USDZExporter exports on the fly from the scene. Scene Viewer cannot do this because it redownloads the URL from the server, so blobs don't work. However, WebXR works fine on Android since it's still the same scene in the browser. Does that help?

**Reply by yoavniran** [reply]
Hi @elalish 
Thanks but I dont think I have any misunderstanding. Perhaps I wasn't clear enough in my first post.

I'm aware that Scene Viewer is Android only and that Quick Look is iOS only. I mentioned QL because the code I pointed to in this library gave me the idea of using the GLTFExporter to create a model binary on the fly. 

Scene Viewer **is capable** of using the Object URL because it has worked a few times when I just run the model through the GLTFExporter without making any texture changes. The problem I have is that whenever I make changes to textures and then run it through the exporter, the SV just breaks.

Unfortunately, I have no error message or way to debug SV to tell why its breaking...

**Reply by elalish** [reply]
No, we don't use GLTFExporter. QL uses USDZExporter, hence why QL works. SV really isn't capable of using an object URL - we just forward the original `src` URL to it, hence why it always shows the unmodified version. If SV is crashing, that's a SV issue, and I'd be happy to forward it if you can give us an easy way to reproduce.

**Reply by yoavniran** [reply]
That'd be great! 
Here's a [test page](https://d8s-test-viewer.s3.us-east-1.amazonaws.com/app/index.html?appMode=qr&model=https://dimensions-art.cloudinary.net/d8s-yoav-test-1/image/upload/v1/dimensions-app/3dmodels/y8dq3d7bl0nydrywpmmf?_a=AQAK+g1&fallback=https://dimensions.cloudinary.com/ar-fallback.html&textures=eyJuYW1lIjoiYXN0cm9uYXV0LmFzdHJvbmF1dF8yIiwiY29uZmlnIjp7Im1hdGVyaWFsUGFyYW1zIjp7ImNvbG9yIjpbImdvbGQiXX19fQ==) opening my Viewer. On an Android device it should show the astronaut with a "open in AR" button.

<img width="500" src="https://github.com/google/model-viewer/assets/1102278/4289fbba-588a-4a90-baf7-29ddd86e0088"/>

Clicking the button should open the Scene Viewer and it crashes every time for me. 
If you could pass that along to the SV folks, it would be awesome.

Thanks!

**Reply by elalish** [reply]
As I say, you can't pass an ObjectURL to Scene Viewer - it means nothing outside of the browser. It shouldn't crash, but it also has no way to work.

**Reply by yoavniran** [reply]
The following [URL](https://d8s-test-viewer.s3.us-east-1.amazonaws.com/app/index.html?appMode=qr&model=https://dimensions-art.cloudinary.net/d8s-yoav-test-1/image/upload/v1/dimensions-app/3dmodels/y8dq3d7bl0nydrywpmmf?_a=AQAK+g1&fallback=https://dimensions.cloudinary.com/ar-fallback.html&textures=) takes the model, exports it with GLTFExporter, and provides the **object URL**, which _does_ load in the Scene Viewer successfully:

<img width="450" src="https://github.com/google/model-viewer/assets/1102278/ca95c031-c769-4b74-a42a-6de983b6241f"/>

Its the same model (astronaut) and same code. Only difference is that there was no change to texture/material...

**Reply by elalish** [reply]
Interesting - news to me that it works that way. @AdrianAtGoogle Can Scene Viewer really handle ObjectURLs now? I don't know what this guy did - it's not MV and his code is obfuscated.
